daisy_workflows/image_test/network/network-tester.py

```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/python
+# Copyright 2018 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import urllib2
+
+
+import utils
+import genips
+
+utils.AptGetInstall(['python-pip'])
+utils.Execute(['pip', 'install', '--upgrade', 'google-api-python-client'])
+
+from googleapiclient import discovery
+import oauth2client.client
+
+
+def GetHostname(addr, timeout=10):
+  URL = "http://%s/hostname" % addr
+  print("Retrieving: %s" % URL)
+  return urllib2.urlopen(URL, None, timeout).read()
+
+
+def TestIPAlias(testee, ip_alias, ip_mask):
+  testee_hostname = GetHostname(testee)
+  for ip in genips.GenIPs(ip_alias, ip_mask):
+    if testee_hostname != GetHostname(ip):
+      raise Exception("Alias hostname should be the same from host machine.")
+
+
+class RequestError(Exception):
+  def __init__(self, name, instance, ip_range):
+    self.name = name
+    self.instance = instance
+    self.ip_range = ip_range
+
+  def __str__(self):
+    return "%s didn't redirect ip %s to instance %s" % (
+        self.name, self.ip_range, self.instance)
+
+
+def TestForwardingRule(MD, instance):
+  target_name = instance + "-target"
+  MD.Wait(MD.AddTargetInstance(target_name, instance))
+  try:
+    rule_name = instance + "-rule"
+    MD.Wait(MD.AddForwardingRule(rule_name, target_name))
+    try:
+      ip = MD.GetForwardingRuleIP(rule_name)
+      if GetHostname(instance) != GetHostname(ip):
+        raise RequestError(rule_name, instance, ip)
+    finally:
+      MD.Wait(MD.DeleteForwardingRule(rule_name))
+  finally:
+    MD.Wait(MD.DeleteTargetInstance(target_name))
+
+
+def main():
+  MM = utils.MetadataManager
+  testee = MM.FetchMetadataDefault('testee')
+
+  # Verify IP aliasing is working
+  TestIPAlias(testee, MM.FetchMetadataDefault('alias_ip'),
+      MM.FetchMetadataDefault('alias_ip_mask'))
+
+  # Ensure routes are added when enabling IP Forwarding
+  compute = utils.GetCompute(discovery, oauth2client.client.GoogleCredentials)
+
+  # (Negative test)
+  # The testee instance has IP Forward disabled, hence it should not accept the
+  # traffic from a different IP even if the router is sending traffic to it.
+  # This was a test to guarantee that behavior.
+  try:
+    TestForwardingRule(MM(compute, testee), testee)
```

Comment:
forgot to raise an exception here: the `TestForwardingRule` should not work on this case. Hang on..

Reply:
done